src/Compilers/CSharp/Portable/Symbols/UnboundGenericType.cs

```diff
@@ -119,4 +119,75 @@ public override int GetHashCode()
                 : Hash.Combine(_name, _errorInfo.Code);
         }
     }
+
+    /// <summary>
+    /// Used for lightweight binding of type constraints. Instead of binding type arguments,
+    /// we'll just use these placeholders instead. That's good enough binding to compute
+    /// <see cref="TypeSymbol.IsValueType"/> with minimal binding.
+    /// </summary>
+    internal sealed class PlaceholderTypeArgumentSymbol : ErrorTypeSymbol
+    {
+        public static ImmutableArray<TypeWithAnnotations> CreateTypeArguments(ImmutableArray<TypeParameterSymbol> typeParameters)
+        {
+            var result = ArrayBuilder<TypeWithAnnotations>.GetInstance();
+            foreach (var typeParameter in typeParameters)
+            {
+                result.Add(TypeWithAnnotations.Create(new PlaceholderTypeArgumentSymbol(typeParameter.Name)));
+            }
+            return result.ToImmutableAndFree();
+        }
+
+        private readonly string _name;
+
+        private PlaceholderTypeArgumentSymbol(string name, TupleExtraData? tupleData = null)
+            : base(tupleData)
+        {
+            _name = name;
+        }
+
+        protected override NamedTypeSymbol WithTupleDataCore(TupleExtraData newData)
+        {
+            return new PlaceholderTypeArgumentSymbol(_name, newData);
+        }
+
+        public override string Name
+        {
+            get
+            {
+                return _name;
+            }
+        }
+
+        internal override bool MangleName
+        {
+            get
+            {
+                Debug.Assert(Arity == 0);
+                return false;
+            }
+        }
+
+        internal override DiagnosticInfo ErrorInfo
+        {
+            get
+            {
+                return new CSDiagnosticInfo(ErrorCode.Void);
```

Comment:
>new CSDiagnosticInfo(ErrorCode.Void); [](start = 23, length = 37)

`null`?

Reply:
The reason I put an error code is to help ensure this symbol doesn't accidentally leak into the full binding, as a redundancy (since it's already an error type).

Reply:
I believe diagnostics with this particular `ErrorCode` are dropped.

---
In reply to: [499842381](https://github.com/dotnet/roslyn/pull/48270#discussion_r499842381) [](ancestors = 499842381)